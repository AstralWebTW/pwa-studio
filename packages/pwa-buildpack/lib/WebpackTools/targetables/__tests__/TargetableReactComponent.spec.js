const TargetableReactComponent = require('../TargetableReactComponent');

const checkboxSource = `import { Component, lazy } from 'react';
export function Checkbox(props) {
    const { classes, fieldState, id, label, message, ...rest } = props;
    const { value: checked } = fieldState;
    const icon = checked ? checkedIcon : uncheckedIcon;

    return (
        <Fragment>
            <label className={classes.root} htmlFor={id}>
                <BasicCheckbox
                    {...rest}
                    className={classes.input}
                    fieldState={fieldState}
                    id={id}
                />
                <span className={classes.icon}>{icon}</span>
                <span className={classes.label}>{label}</span>
            </label>
            <Message fieldState={fieldState}>{message}</Message>
        </Fragment>
    );
}
`;

const fileToTransform = '/path/to/Checkbox.js';
const CheckboxModule = new TargetableReactComponent(fileToTransform, () => {});
const latestTransform = () =>
    CheckboxModule._queuedTransforms[
        CheckboxModule._queuedTransforms.length - 1
    ];

test('.addLazyReactComponentImport()', () => {
    CheckboxModule.addLazyReactComponentImport(
        './path/to/dynamic/component',
        'Blerg'
    );
    expect(latestTransform()).toMatchSnapshot();
});
describe('JSX manipulation', () => {
    it('.appendJSX()', () => {
        CheckboxModule.appendJSX(
            'span className={classes.icon}',
            '<AnotherEmoji />'
        );
        expect(latestTransform()).toMatchSnapshot();
    });
    it('.insertAfterJSX()', () => {
        CheckboxModule.insertAfterJSX('<Message>', '<Rakim />');
        expect(latestTransform()).toMatchSnapshot();
    });
    it('.insertBeforeJSX()', () => {
        CheckboxModule.insertBeforeJSX('<Message>', '<EricB and={and} />');
        expect(latestTransform()).toMatchSnapshot();
    });
    it('.prependJSX()', () => {
        CheckboxModule.prependJSX(
            '<span className={classes.icon} />',
            '<AnEmoji/>'
        );
        expect(latestTransform()).toMatchSnapshot();
    });
    it('.removeJSX()', () => {
        CheckboxModule.removeJSX('AnotherEmoji');
        expect(latestTransform()).toMatchSnapshot();
    });
    it('.removeJSXProps()', () => {
        CheckboxModule.removeJSXProps('BasicCheckbox', ['fieldState', 'id']);
        expect(latestTransform()).toMatchSnapshot();
    });
    it('.replaceJSX()', () => {
        CheckboxModule.replaceJSX(
            'span className={classes.label}',
            '<i>where is label oh no</i>'
        );
        expect(latestTransform()).toMatchSnapshot();
    });
    it('.setJSXProps()', () => {
        CheckboxModule.setJSXProps('BasicCheckbox', {
            key: '{NUMBER}',
            className: '{classes.wildin}'
        });
        expect(latestTransform()).toMatchSnapshot();
    });
    it('.surroundJSX()', () => {
        CheckboxModule.surroundJSX('label', '<fieldset></fieldset>');
        expect(latestTransform()).toMatchSnapshot();
    });

    it('really ties the room together', async () => {
        const babel = require('@babel/core');
        const { code } = await babel.transformAsync(checkboxSource, {
            plugins: [
                [
                    require.resolve('../BabelModifyJSXPlugin/plugin.js'),
                    {
                        requestsByFile: {
                            [fileToTransform]: CheckboxModule.flush().filter(
                                r =>
                                    r.transformModule.includes(
                                        'BabelModifyJSXPlugin'
                                    )
                            )
                        }
                    }
                ]
            ],
            filename: fileToTransform
        });
        expect(code).toMatchSnapshot();
    });
});
