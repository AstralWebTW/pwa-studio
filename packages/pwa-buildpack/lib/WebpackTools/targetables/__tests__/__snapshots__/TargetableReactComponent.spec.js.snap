// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`.addLazyReactComponentImport() 1`] = `
Object {
  "fileToTransform": "/path/to/Checkbox.js",
  "options": Object {
    "after": "{ lazy as reactLazy } from 'react';
",
    "insert": "const DynamicBlerg$1 = reactLazy(() => import('./path/to/dynamic/component'));
",
  },
  "transformModule": "@magento/pwa-buildpack/lib/WebpackTools/loaders/splice-source-loader",
  "type": "source",
}
`;

exports[`JSX manipulation .appendJSX() 1`] = `
Object {
  "fileToTransform": "/path/to/Checkbox.js",
  "options": Object {
    "element": "span className={classes.icon}",
    "operation": "append",
    "params": Object {
      "jsx": "<AnotherEmoji />",
    },
  },
  "transformModule": "/Volumes/CaseSensitive/pwa-studio/packages/pwa-buildpack/lib/WebpackTools/targetables/BabelModifyJSXPlugin/index.js",
  "type": "babel",
}
`;

exports[`JSX manipulation .insertAfterJSX() 1`] = `
Object {
  "fileToTransform": "/path/to/Checkbox.js",
  "options": Object {
    "element": "<Message>",
    "operation": "insertAfter",
    "params": Object {
      "jsx": "<Rakim />",
    },
  },
  "transformModule": "/Volumes/CaseSensitive/pwa-studio/packages/pwa-buildpack/lib/WebpackTools/targetables/BabelModifyJSXPlugin/index.js",
  "type": "babel",
}
`;

exports[`JSX manipulation .insertBeforeJSX() 1`] = `
Object {
  "fileToTransform": "/path/to/Checkbox.js",
  "options": Object {
    "element": "<Message>",
    "operation": "insertBefore",
    "params": Object {
      "jsx": "<EricB and={and} />",
    },
  },
  "transformModule": "/Volumes/CaseSensitive/pwa-studio/packages/pwa-buildpack/lib/WebpackTools/targetables/BabelModifyJSXPlugin/index.js",
  "type": "babel",
}
`;

exports[`JSX manipulation .prependJSX() 1`] = `
Object {
  "fileToTransform": "/path/to/Checkbox.js",
  "options": Object {
    "element": "<span className={classes.icon} />",
    "operation": "prepend",
    "params": Object {
      "jsx": "<AnEmoji/>",
    },
  },
  "transformModule": "/Volumes/CaseSensitive/pwa-studio/packages/pwa-buildpack/lib/WebpackTools/targetables/BabelModifyJSXPlugin/index.js",
  "type": "babel",
}
`;

exports[`JSX manipulation .removeJSX() 1`] = `
Object {
  "fileToTransform": "/path/to/Checkbox.js",
  "options": Object {
    "element": "AnotherEmoji",
    "operation": "remove",
    "params": undefined,
  },
  "transformModule": "/Volumes/CaseSensitive/pwa-studio/packages/pwa-buildpack/lib/WebpackTools/targetables/BabelModifyJSXPlugin/index.js",
  "type": "babel",
}
`;

exports[`JSX manipulation .removeJSXProps() 1`] = `
Object {
  "fileToTransform": "/path/to/Checkbox.js",
  "options": Object {
    "element": "BasicCheckbox",
    "operation": "removeProps",
    "params": Object {
      "props": Array [
        "fieldState",
        "id",
      ],
    },
  },
  "transformModule": "/Volumes/CaseSensitive/pwa-studio/packages/pwa-buildpack/lib/WebpackTools/targetables/BabelModifyJSXPlugin/index.js",
  "type": "babel",
}
`;

exports[`JSX manipulation .replaceJSX() 1`] = `
Object {
  "fileToTransform": "/path/to/Checkbox.js",
  "options": Object {
    "element": "span className={classes.label}",
    "operation": "replace",
    "params": Object {
      "jsx": "<i>where is label oh no</i>",
    },
  },
  "transformModule": "/Volumes/CaseSensitive/pwa-studio/packages/pwa-buildpack/lib/WebpackTools/targetables/BabelModifyJSXPlugin/index.js",
  "type": "babel",
}
`;

exports[`JSX manipulation .setJSXProps() 1`] = `
Object {
  "fileToTransform": "/path/to/Checkbox.js",
  "options": Object {
    "element": "BasicCheckbox",
    "operation": "setProps",
    "params": Object {
      "props": Object {
        "className": "{classes.wildin}",
        "key": "{NUMBER}",
      },
    },
  },
  "transformModule": "/Volumes/CaseSensitive/pwa-studio/packages/pwa-buildpack/lib/WebpackTools/targetables/BabelModifyJSXPlugin/index.js",
  "type": "babel",
}
`;

exports[`JSX manipulation .surroundJSX() 1`] = `
Object {
  "fileToTransform": "/path/to/Checkbox.js",
  "options": Object {
    "element": "label",
    "operation": "surround",
    "params": Object {
      "jsx": "<fieldset></fieldset>",
    },
  },
  "transformModule": "/Volumes/CaseSensitive/pwa-studio/packages/pwa-buildpack/lib/WebpackTools/targetables/BabelModifyJSXPlugin/index.js",
  "type": "babel",
}
`;

exports[`JSX manipulation really ties the room together 1`] = `
"\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.Checkbox = Checkbox;

var _react = require(\\"react\\");

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function Checkbox(props) {
  const {
    classes,
    fieldState,
    id,
    label,
    message
  } = props,
        rest = _objectWithoutProperties(props, [\\"classes\\", \\"fieldState\\", \\"id\\", \\"label\\", \\"message\\"]);

  const {
    value: checked
  } = fieldState;
  const icon = checked ? checkedIcon : uncheckedIcon;
  return React.createElement(Fragment, null, React.createElement(\\"fieldset\\", null, React.createElement(\\"label\\", {
    className: classes.root,
    htmlFor: id
  }, React.createElement(BasicCheckbox, _extends({}, rest, {
    className: classes.wildin,
    key: NUMBER
  })), React.createElement(\\"span\\", {
    className: classes.icon
  }, React.createElement(AnEmoji, null), icon), React.createElement(\\"i\\", null, \\"where is label oh no\\"))), React.createElement(EricB, {
    and: and
  }), React.createElement(Message, {
    fieldState: fieldState
  }, message), React.createElement(Rakim, null));
}"
`;
